### PR TITLE
pkg tlsf: make it compile with -pedantic

### LINKED ISSUE #4451: pkg: tlsf: not building with -pedantic
gcc with pedantic doesn't like the @Kijewski's patch:
`/home/oleg/git/RIOT/examples/ccn-lite-relay/bin/native/tlsf-src/tlsf.c:161:6: error: ISO C forbids specifying range of elements to initialize [-Werror=pedantic]
   [0 ... FL_INDEX_COUNT-1] = {`

@Kijewski, would you object reverting this part of the patch?

The corresponding code looks like:
```C
/* The TLSF control structure. */
typedef struct control_t
{
    /* Empty lists point at this block to indicate they are free. */
    block_header_t block_null;

    /* Bitmaps for free lists. */
    unsigned int fl_bitmap;
    unsigned int sl_bitmap[FL_INDEX_COUNT];

    /* Head of free lists. */
    block_header_t* blocks[FL_INDEX_COUNT][SL_INDEX_COUNT];
} control_t;

static control_t control __attribute__ ((aligned(ALIGN_SIZE))) = {
    .block_null = {
        .next_free = &control.block_null,
        .prev_free = &control.block_null,
    },
    .fl_bitmap = 0,
    .sl_bitmap = {0},
    .blocks = {
        [0 ... FL_INDEX_COUNT-1] = {
            [0 ... SL_INDEX_COUNT-1] = &control.block_null
        }
    }
};
```

### PR DESCRIPTION
**Pull request by OlegHahm**
Fixes #4451 by re-introducing the init function and updates using functions accordingly.

### CODE REVIEW

**Comment by OlegHahm** [comment]
btw. I verified that no malloc or calloc is called before main on the SAMR21-xpro, which should in turn be true for at least all Cortex-M platforms.

**Comment by OlegHahm** [comment]
Checked also for nano-specs being disabled.

**Comment by kaspar030** [comment]
> btw. I verified that no malloc or calloc is called before main on the SAMR21-xpro, which should in turn be true for at least all Cortex-M platforms.

Is that also true when auto_init functions output something? didn't at some point nanolib without nanospecs call malloc when printing?

**Comment by OlegHahm** [comment]
I'll give a try.

**Comment by OlegHahm** [comment]
Checked. Seems not to be the case.

How I tested: I applied
```diff
diff --git a/boards/Makefile.include.cortexm_common b/boards/Makefile.include.cortexm_common
index 93073ff..37de995 100644
--- a/boards/Makefile.include.cortexm_common
+++ b/boards/Makefile.include.cortexm_common
@@ -47,10 +47,5 @@ export LINKFLAGS += -Wl,--gc-sections
 # Import all toolchain settings
 include $(RIOTBOARD)/Makefile.include.$(TOOLCHAIN)
 
-# use the nano-specs of Newlib when available
-ifeq ($(shell $(LINK) -specs=nano.specs -E - 2>/dev/null >/dev/null </dev/null ; echo $$?),0)
-export LINKFLAGS += -specs=nano.specs -lc -lnosys
-endif
-
 # export board specific includes to the global includes-listing
 export INCLUDES += -I$(RIOTBOARD)/$(BOARD)/include
```
then compiled, flashed, and debugged setting a breakpoint to calloc and one to malloc. It didn't fire before it enters main. The breakpoint is hit as soon as I call some CCNL function.

**Comment by kaspar030** [comment]
Why don't you use auto_init to initialize tlsf?

**Comment by OlegHahm** [comment]
a) that's IMO another PR
b) we may consider initializing it even earlier in case some weird libc or bsp wants to use malloc
c) the more interesting question for auto initialization is how to get the default values

ad c) @haukepetersen proposed to get the top of the stack and end of RAM pointer from, e.g., the linkerscript, and then just reserver the full space (aka heap) for TLSF. IMO that seems reasonable.

**Comment by OlegHahm** [comment]
I realized that this patch is actually incorrect. The control data will be stored on the stack. I wonder if anyone has a preference whether to store it there or rather in the same memory as provided during initialization. Anyone with a preference?

**Comment by OlegHahm** [comment]
Ok, fixed that. I decided to put the memory for the control block inside the memory block that is given during initialization. Rationale: this increase the flexibility. One can still pass a buffer from the stack to this function and pass additional pools to TLSF using the `add_pool()` function.

In order to ease reviewing the patch, I opened a Git repository: https://github.com/OlegHahm/tlsf

File: examples/ccn-lite-relay/main.c [hunk 1]
```diff
@@ -33,17 +29,13 @@
 #define MAIN_QUEUE_SIZE     (8)
 static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
 
-#ifndef BOARD_NATIVE
-    /* some buffer for the heap */
-#   define TLSF_BUFFER     (10240 / sizeof(uint32_t))
-    static uint32_t _tlsf_heap[TLSF_BUFFER];
-#endif
+/* some buffer for the heap */
+#define TLSF_BUFFER     (10240 / sizeof(uint32_t))
```
Comment (haukepetersen) [inline examples/ccn-lite-relay/main.c]: where does this magic number come from? Any reason for such a weird size?

Comment (OlegHahm) [inline examples/ccn-lite-relay/main.c]: Hm, not really, just tried to find a sufficient big number that would work on all supported platforms and though 10kB would be a nice number.

**Comment by haukepetersen** [comment]
The changes look sane to me, and the adapted example work fine (they compile + shell interaction works) on native and on the `samr21`. 

@OlegHahm: could you please open an issue for moving the initialization to auto_init or even the cpu code and using some memory markers to use the entire free memory so we won't forget about this? And now that I think of it: maybe even a second issue to print the actual used heap space when using the `ps` command?

Otherwise: ACK

**Comment by authmillenon** [comment]
> And now that I think of it: maybe even a second issue to print the actual used heap space when using the `ps` command?

What does heap usage has to do with process information?

**Comment by OlegHahm** [comment]
> could you please open an issue for moving the initialization to auto_init or even the cpu code and using some memory markers to use the entire free memory so we won't forget about this? 

Will do. Actually, I have already a branch working on this.

> And now that I think of it: maybe even a second issue to print the actual used heap space when using the ps command?

I'm also already working no this. But in a dedicated command. (I agree with @authmillenon, it's rather `free` than `ps`.)

**Comment by OlegHahm** [comment]
Travis is green, @haukepetersen acked, and follow-up issues (#4490 and #4491) are open.

**Pull request merged**